lib/vimgolf/lib/vimgolf/keylog.rb

```diff
@@ -1,258 +1,267 @@
-              when "k6"; "<F6>"
-              when "k7"; "<F7>"
-              when "k8"; "<F8>"
-              when "k9"; "<F9>"
-              when "k;"; "<F10>"
-              when "F1"; "<F11>"
-              when "F2"; "<F12>"
-              when "F3"; "<F13>"
-              when "F4"; "<F14>"
-              when "F5"; "<F15>"
-              when "F6"; "<F16>"
-              when "F7"; "<F17>"
-              when "F8"; "<F18>"
-              when "F9"; "<F19>"
-
-              when "%1"; "<Help>"
-              when "&8"; "<Undo>"
-              when "#2"; "<S-Home>"
-              when "*7"; "<S-End>"
-              when "K1"; "<kHome>"
-              when "K4"; "<kEnd>"
-              when "K3"; "<kPageUp>"
-              when "K5"; "<kPageDown>"
-              when "K6"; "<kPlus>"
-              when "K7"; "<kMinus>"
-              when "K8"; "<kDivide>"
-              when "K9"; "<kMultiply>"
-              when "KA"; "<kEnter>"
-              when "KB"; "<kPoint>"
-              when "KC"; "<k0>"
-              when "KD"; "<k1>"
-              when "KE"; "<k2>"
-              when "KF"; "<k3>"
-              when "KG"; "<k4>"
-              when "KH"; "<k5>"
-              when "KI"; "<k6>"
-              when "KJ"; "<k7>"
-              when "KK"; "<k8>"
-              when "KL"; "<k9>"
-
-              when "kP"; "<PageUp>"
-              when "kN"; "<PageDown>"
-              when "kh"; "<Home>"
-              when "@7"; "<End>"
-              when "kI"; "<Insert>"
-              when "kD"; "<Del>"
-              when "kb"; "<BS>"
+      # A Vim keycode is either a single byte, or a 3-byte sequence starting
+      # with 0x80.
+      while (c = scanner.get_byte)
+        n = c.ord
+        if n == 0x80
+          b2, b3 = scanner.get_byte, scanner.get_byte
+          if b2 == "\xfd" && b3 >= "\x38" && @time > @@sniff_date
+            # Should we account for KE_SNIFF removal?
+            b3 = (b3.ord + 1).chr
+          end
+          code = @@kc_mbyte[b2+b3]
+          yield code if code # ignore "nil" keystrokes (like window focus)
+        else
+          yield @@kc_1byte[n]
+        end
+      end
+    end
 
-              when "ku"; "<Up>"
-              when "kd"; "<Down>"
-              when "kl"; "<Left>"
-              when "kr"; "<Right>"
-              when "#4"; "<S-Left>"
-              when "%i"; "<S-Right>"
+    # Quick lookup array for single-byte keycodes
+    @@kc_1byte = []
+    (0..255).each {|n| @@kc_1byte.push("<%#04x>" % n)} # Fallback for non-ASCII
+    (1..127).each {|n| @@kc_1byte[n] = "<C-#{(n ^ 0x40).chr}>"}
+    (32..126).each {|c| @@kc_1byte[c] = c.chr } # Printing chars
+    @@kc_1byte[0x1b] = "<Esc>" # Special names for a few control chars
+    @@kc_1byte[0x0d] = "<CR>"
+    @@kc_1byte[0x0a] = "<NL>"
+    @@kc_1byte[0x09] = "<Tab>"
 
-              when "kB"; "<S-Tab>"
-              when "\xffX"; "<C-@>"
+    # After this date, assume KE_SNIFF is removed
+    @@sniff_date = Time.utc(2016, 4)
```

Comment:
Any reason why we can't make this a const? 

Ditto for other @@'s.. we can create and freeze the hashes? If we never change any of these, perhaps that's a better route?